tests/test_openclhpp.cpp

```diff
@@ -3626,7 +3630,86 @@ void testDevice_GetInfo_CLDeviceName()
+    cl_platform_id id, cl_platform_info param_name, size_t param_value_size,
+    void *param_value, size_t *param_value_size_ret, int num_calls) {
+    (void)num_calls;
+    const char *version = {"OpenCL 1.1 Mock"};
+    size_t bytes = strlen(version) + 1;
+    TEST_ASSERT_NOT_NULL(id);
+    TEST_ASSERT_EQUAL_PTR(make_platform_id(0), id);
+    TEST_ASSERT_EQUAL_HEX(CL_PLATFORM_VERSION, param_name);
+    TEST_ASSERT(param_value == nullptr || param_value_size >= bytes);
+    if (param_value_size_ret != nullptr)
+        *param_value_size_ret = bytes;
+    if (param_value != nullptr)
+        strcpy((char *) param_value, version);
+    
+    return CL_SUCCESS;
+}
+
+static cl_int clGetDeviceInfo_testDevice_createSubDevices(
+    cl_device_id id, cl_device_info param_name, size_t param_value_size,
+    void *param_value, size_t *param_value_size_ret, int num_calls) {
+    (void)id;
+    (void)num_calls;
+
+    TEST_ASSERT_EQUAL_HEX(CL_DEVICE_PLATFORM, param_name);
+    TEST_ASSERT(param_value == nullptr || param_value_size >= sizeof(cl_device_id));
+    if (param_value_size_ret != nullptr)
+        *param_value_size_ret = sizeof(cl_device_id);
+    if (param_value != nullptr)
+        *(cl_platform_id *)param_value = make_platform_id(0);
+    return CL_SUCCESS;
+    
+}
+
+static cl_int clCreateSubDevices_testDevice_createSubDevices(
+    cl_device_id in_device, const cl_device_partition_property *properties,
+    cl_uint num_devices, cl_device_id *out_devices, cl_uint *num_devices_ret,
+    int num) {
+    return 0;
+}
+
+static cl_int clCreateSubDevicesEXT_testDevice_createSubDevices(
+    cl_device_id device_in, const cl_device_partition_property_ext *properties,
+    cl_uint n, cl_device_id *out_devices, cl_uint *num, int num_calls) {
+  cl_int ret = CL_SUCCESS;
+  if (device_in == make_device_id(0)) {
+    return CL_SUCCESS;
+  } else if (device_in == make_device_id(1) && nullptr == num) {
+    return CL_DEVICE_NOT_FOUND;
+  } else {
+    return CL_DEVICE_NOT_FOUND;
+  }
+}
+
+void testDevice_createSubDevices() {
+  const cl_device_partition_property_ext properties =
+      CL_DEVICE_PARTITION_TYPES_EXT;
+  std::vector<cl::Device> devices;
+  cl_platform_id platform = make_platform_id(0);
+
+  clGetPlatformInfo_StubWithCallback(
+      clGetPlatformInfo_testDevice_createSubDevices);
+  clGetDeviceInfo_StubWithCallback(clGetDeviceInfo_testDevice_createSubDevices);
+  clCreateSubDevices_StubWithCallback(
+      clCreateSubDevices_testDevice_createSubDevices);
+
+  clCreateSubDevicesEXT_StubWithCallback(
+      clCreateSubDevicesEXT_testDevice_createSubDevices);
+
+  cl_int ret = devicePool[0].createSubDevices(&properties, &devices);
+  TEST_ASSERT_EQUAL(CL_SUCCESS, ret);
+  ret = devicePool[1].createSubDevices(&properties, &devices);
+  TEST_ASSERT_EQUAL(CL_DEVICE_NOT_FOUND, ret);
+  ret = devicePool[2].createSubDevices(&properties, &devices);
+  TEST_ASSERT_EQUAL(CL_DEVICE_NOT_FOUND, ret);
```

Comment:
These tests that return error codes will all "fail" for the `CL_HPP_ENABLE_EXCEPTIONS` tests, because the error conditions will cause exceptions to be thrown.  Options:

1. They should only be run with `CL_HPP_ENABLE_EXCEPTIONS` is not defined.
2. They should be rewritten so they do not return error codes.

Reply:
We need some "fails" to cover all options. CL_HPP_ENABLE_EXCEPTIONS added